### PR TITLE
:FC-776: :sparkles: Invocations by Skill

### PR DESCRIPTION
**Pull request by agalagade**
FaaS and Controller Bot feature that allows users to set restrictions on which skill actions invoke their lambdas in a self-service manner.

### CODE REVIEW

File: test/isImplementedCache.test.ts [hunk 1]
```diff
@@ -14,22 +14,28 @@ describe('ImplementedCache', () => {
         isImplementedCache.add(event, true);
 
         expect(isImplementedCache.get(event)).toBeDefined();
-        expect(isImplementedCache.get(event)?.isImplemented).toBeTrue();
+        expect(isImplementedCache.get(event)?.isImplemented).toBe(true);
```
Comment (Templum) [inline test/isImplementedCache.test.ts:17]: Why was this changed ?

File: test/isImplementedCache.test.ts [hunk 2]
```diff
@@ -14,22 +14,28 @@ describe('ImplementedCache', () => {
...
       }
     );
     it('should overwrite a specific event that already exists in cache', () => {
       isImplementedCache.add(event, true);
-      isImplementedCache.add(event, false);
+      isImplementedCache.add(event, true);
```
Comment (Templum) [inline test/isImplementedCache.test.ts:22]: why is this not false ?

File: test/isImplementedCache.test.ts [hunk 3]
```diff
@@ -14,22 +14,28 @@ describe('ImplementedCache', () => {
...
 
       expect(isImplementedCache.get(event)).toBeDefined();
-      expect(isImplementedCache.get(event)?.isImplemented).toBeFalse();
+      expect(isImplementedCache.get(event)?.isImplemented).toBe(false);
```
Comment (Templum) [inline test/isImplementedCache.test.ts:25]: How can that be false, if you set true ?

File: test/isImplementedCache.test.ts [hunk 4]
```diff
@@ -14,22 +14,28 @@ describe('ImplementedCache', () => {
...
       isImplementedCache = new IsImplementedCache(-2);
       isImplementedCache.add(event, true);
 
       expect(isImplementedCache.get(event)).toBeUndefined();
     });
+    it('should return an event that has skillId', () => {
+      isImplementedCache.add(event, true, 'smapleSkill');
+
+      expect(isImplementedCache.get(event)).toBeDefined();
+      expect(isImplementedCache.get(event)?.skillId).toBeDefined();
```
Comment (Templum) [inline test/isImplementedCache.test.ts:37]: Not good, you should check for the type as well.

**Review by Templum — CHANGES_REQUESTED** [review]
I see not adjustment to the isImplemented rest call.

File: test/isImplementedCache.test.ts (hunk 1 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts:17]: toBeTrue was flagging as an error on my Editor that is why I changed it. i have reverted my changes

File: test/isImplementedCache.test.ts (hunk 2 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts:22]: resolved

File: test/isImplementedCache.test.ts (hunk 3 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts:25]: resolved

File: test/isImplementedCache.test.ts (hunk 4 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts:37]: updated

File: src/helper/isImplementedCache.ts [hunk 5]
```diff
@@ -20,7 +21,7 @@ export class IsImplementedCache {
     return undefined;
   }
 
-  add(eventName: string, isImplemented: boolean): void {
+  add(eventName: string, isImplemented: boolean, skillId?: string): void {
```
Comment (Templum) [inline src/helper/isImplementedCache.ts:24]: Right now the code only adds the skillId as part of the entry (but does not use it at all afterwards). Although that should be relevant.

**Review by Templum — CHANGES_REQUESTED** [review]

File: src/helper/isImplementedCache.ts (hunk 5 above)
Comment (agalagade) [reply src/helper/isImplementedCache.ts:31]: added changes

File: src/helper/isImplementedCache.ts [hunk 6]
```diff
@@ -10,10 +10,17 @@ export class IsImplementedCache {
...
+    let event: ImplementedEvent | undefined;
+    if (skillId) {
+      event = this.cache.find(e => {
+        return e.name === eventName && e.skillId === skillId;
+      });
+    } else {
+      event = this.cache.find(e => {
+        return e.name === eventName;
+      });
+    }
```
Comment (Templum) [inline src/helper/isImplementedCache.ts:23]: This part can be handled in one line of code, avoiding the if-else statement.

```javascript
event = this.cache.find(e => { return skillId ? e.name === eventName && e.skillId === skillId: e.name === eventName });
```

File: test/isImplementedCache.test.ts [hunk 7]
```diff
@@ -33,9 +33,14 @@ describe('ImplementedCache', () => {
     it('should return an event that has skillId', () => {
       isImplementedCache.add(event, true, 'smapleSkill');
 
-      expect(isImplementedCache.get(event)).toBeDefined();
-      expect(isImplementedCache.get(event)?.skillId).toBeDefined();
-      expect(isImplementedCache.get(event)?.skillId).toBeString();
+      expect(isImplementedCache.get(event,'smapleSkill')).toBeDefined();
+      expect(isImplementedCache.get(event,'smapleSkill')?.skillId).toBeDefined();
```
Comment (Templum) [inline test/isImplementedCache.test.ts:37]: unnecessary based on next line

File: test/isImplementedCache.test.ts [hunk 8]
```diff
@@ -33,9 +33,14 @@ describe('ImplementedCache', () => {
...
+      expect(isImplementedCache.get(event,'smapleSkill')?.skillId).toBeString();
```
Comment (Templum) [inline test/isImplementedCache.test.ts:38]: You could actually go so far to make sure that skill is equal to `smapleSkill`

**Review by Templum — CHANGES_REQUESTED** [review]

File: src/helper/isImplementedCache.ts (hunk 6 above)
Comment (agalagade) [reply src/helper/isImplementedCache.ts:23]: resolved

File: test/isImplementedCache.test.ts (hunk 7 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts:37]: updated

File: test/isImplementedCache.test.ts (hunk 8 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts:38]: updated

File: test/isImplementedCache.test.ts [hunk 9]
```diff
@@ -30,6 +30,18 @@ describe('ImplementedCache', () => {
 
       expect(isImplementedCache.get(event)).toBeUndefined();
     });
+    it('should return an event that has skillId', () => {
```
Comment (Templum) [inline test/isImplementedCache.test.ts:33]:
```suggestion
    it('should return the event with the provided skillId', () => {
```

File: test/isImplementedCache.test.ts [hunk 10]
```diff
@@ -30,6 +30,18 @@ describe('ImplementedCache', () => {
...
+      isImplementedCache.add(event, true, 'sampleSkill');
+
+      expect(isImplementedCache.get(event,'sampleSkill')).toBeDefined();
+      expect(isImplementedCache.get(event,'sampleSkill')?.skillId).toBeString();
+      expect(isImplementedCache.get(event,'sampleSkill')?.skillId).toEqual('sampleSkill');
+    });
+    it('should not return an event when filtering by wrong skillId', () => {
```
Comment (Templum) [inline test/isImplementedCache.test.ts:40]:
```suggestion
    it('should not return an undefined if no entry for provided skillId was found', () => {
```

**Review by Templum — CHANGES_REQUESTED** [review]
Only two minor comments left. Else it looks good to me.

Please make sure to also bump the version in the `package.json` (as you need to manually release it later).

File: test/isImplementedCache.test.ts (hunk 10 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts]: updated

File: test/isImplementedCache.test.ts (hunk 9 above)
Comment (agalagade) [reply test/isImplementedCache.test.ts]: updated

File: package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "liveperson-functions-client",
-  "version": "1.0.11",
+  "version": "1.0.12",
```
Comment (Templum) [inline package.json:3]: It's not a patch it is actually a none breaking change, that introduces a new feature

**Review by Templum — DISMISSED** [review]

File: package.json (hunk 11 above)
Comment (agalagade) [reply package.json]: updated

**Review by Templum — APPROVED** [review]
LGTM
